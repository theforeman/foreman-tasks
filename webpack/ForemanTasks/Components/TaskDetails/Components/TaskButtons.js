import React from 'react';
import PropTypes from 'prop-types';
import { Col, Button } from 'patternfly-react';
import { useForemanModal } from 'foremanReact/components/ForemanModal/ForemanModalHooks';
import { translate as __ } from 'foremanReact/common/I18n';
import {
  UNLOCK_MODAL,
  FORCE_UNLOCK_MODAL,
} from '../../TaskActions/TaskActionsConstants';

export const TaskButtons = ({
  canEdit,
  dynflowEnableConsole,
  taskReloadStart,
  taskProgressToggle,
  taskReload,
  externalId,
  id,
  action,
  state,
  resumable,
  cancellable,
  hasSubTasks,
  parentTask,
  cancelTaskRequest,
  resumeTaskRequest,
}) => {
  const unlockModalActions = useForemanModal({
    id: UNLOCK_MODAL,
  });
  const forceUnlockModalActions = useForemanModal({
    id: FORCE_UNLOCK_MODAL,
  });
  const editActionsTitle = canEdit
    ? undefined
    : __('You do not have permission');
  const dynflowTitle = dynflowEnableConsole
    ? undefined
    : `dynflow_enable_console ${__('Setting is off')}`;

  return (
    <Col xs={12}>
      <Button
        className="reload-button"
        bsSize="small"
        onClick={taskProgressToggle}
      >
        <span
          className={`glyphicon glyphicon-refresh ${taskReload ? 'spin' : ''}`}
        />
        {taskReload ? __('Stop auto-reloading') : __('Start auto-reloading')}
      </Button>
      <Button
        className="dynflow-button"
        bsSize="small"
        href={`/foreman_tasks/dynflow/${externalId}`}
        disabled={!dynflowEnableConsole}
        rel="noopener noreferrer"
        target="_blank"
      >
        <span title={dynflowTitle} data-original-title={dynflowTitle}>
          {__('Dynflow console')}
        </span>
      </Button>
      <Button
        className="resume-button"
        bsSize="small"
        title={editActionsTitle}
        data-original-title={editActionsTitle}
        disabled={!canEdit || !resumable}
        onClick={() => {
          if (!taskReload) {
            taskReloadStart(id);
          }
          resumeTaskRequest(id, action);
        }}
      >
        {__('Resume')}
      </Button>
      <Button
        className="cancel-button"
        bsSize="small"
        title={editActionsTitle}
        data-original-title={editActionsTitle}
        disabled={!canEdit || !cancellable}
        onClick={() => {
          if (!taskReload) {
            taskReloadStart(id);
          }
          cancelTaskRequest(id, action);
        }}
      >
        {__('Cancel')}
      </Button>
      {parentTask && (
        <Button
          className="parent-button"
          bsSize="small"
          href={`/foreman_tasks/tasks/${parentTask}`}
        >
          {__('Parent task')}
        </Button>
      )}
      {hasSubTasks && (
        <Button
          className="subtask-button"
          bsSize="small"
          href={`/foreman_tasks/tasks/${id}/sub_tasks`}
        >
          {__('Sub tasks')}
        </Button>
      )}
      <Button
        className="unlock-button"
        bsSize="small"
        disabled={!canEdit || state !== 'paused'}
        onClick={unlockModalActions.setModalOpen}
        title={editActionsTitle}
        data-original-title={editActionsTitle}
      >
        {__('Unlock')}
      </Button>
      <Button
        className="force-unlock-button"
        bsSize="small"
        disabled={!canEdit || state === 'stopped'}
        onClick={forceUnlockModalActions.setModalOpen}
        title={editActionsTitle}
        data-original-title={editActionsTitle}
      >
        {__('Force Unlock')}
      </Button>
    </Col>
  );
};

TaskButtons.propTypes = {
  canEdit: PropTypes.bool,
  dynflowEnableConsole: PropTypes.bool,
  taskReloadStart: PropTypes.func.isRequired,
  taskProgressToggle: PropTypes.func.isRequired,
  taskReload: PropTypes.bool,
  externalId: PropTypes.string,
  id: PropTypes.string.isRequired,
  action: PropTypes.string,
  state: PropTypes.string,
  resumable: PropTypes.bool,
  cancellable: PropTypes.bool,
  hasSubTasks: PropTypes.bool,
  parentTask: PropTypes.string,
  cancelTaskRequest: PropTypes.func,
  resumeTaskRequest: PropTypes.func,
};

TaskButtons.defaultProps = {
  canEdit: false,
  dynflowEnableConsole: false,
  taskReload: false,
  externalId: '',
  action: '',
  state: '',
  resumable: false,
  cancellable: false,
  hasSubTasks: false,
  parentTask: '',
  cancelTaskRequest: () => null,
  resumeTaskRequest: () => null,
};
