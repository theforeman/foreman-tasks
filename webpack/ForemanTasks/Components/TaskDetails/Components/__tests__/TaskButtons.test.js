import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { STATUS } from 'foremanReact/constants';
import { TaskButtons } from '../TaskButtons';

const setUnlockModalOpen = jest.fn();
const setForceUnlockModalOpen = jest.fn();

const defaultProps = {
  id: 'test',
  taskReloadStart: jest.fn(),
  taskProgressToggle: jest.fn(),
  cancelTaskRequest: jest.fn(),
  resumeTaskRequest: jest.fn(),
  setUnlockModalOpen,
  setForceUnlockModalOpen,
};

describe('TaskButtons', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('rendering', () => {
    it('renders reload button with correct text when taskReload is false', () => {
      render(<TaskButtons {...defaultProps} taskReload={false} />);
      expect(
        screen.getByRole('button', { name: /start auto-reloading/i })
      ).toBeInTheDocument();
      expect(
        screen.queryByRole('button', { name: /stop auto-reloading/i })
      ).not.toBeInTheDocument();
    });

    it('renders reload button with correct text when taskReload is true', () => {
      render(<TaskButtons {...defaultProps} taskReload />);
      expect(
        screen.getByRole('button', { name: /stop auto-reloading/i })
      ).toBeInTheDocument();
      expect(
        screen.queryByRole('button', { name: /start auto-reloading/i })
      ).not.toBeInTheDocument();
    });

    it('renders dynflow console link with correct href when externalId is provided', () => {
      render(<TaskButtons {...defaultProps} externalId="external-123" />);
      const dynflowLink = screen.getByRole('link', {
        name: /dynflow console/i,
      });
      expect(dynflowLink).toHaveAttribute(
        'href',
        '/foreman_tasks/dynflow/external-123'
      );
      expect(dynflowLink).toHaveAttribute('target', '_blank');
      expect(dynflowLink).toHaveAttribute('rel', 'noopener noreferrer');
    });

    it('disables dynflow console link when dynflowEnableConsole is false', () => {
      render(<TaskButtons {...defaultProps} dynflowEnableConsole={false} />);
      const dynflowLink = screen.getByRole('link', {
        name: /dynflow console/i,
      });
      expect(dynflowLink).toHaveClass('disabled');
    });

    it('enables dynflow console link when dynflowEnableConsole is true', () => {
      render(
        <TaskButtons
          {...defaultProps}
          dynflowEnableConsole
          externalId="external-123"
        />
      );
      const dynflowLink = screen.getByRole('link', {
        name: /dynflow console/i,
      });
      expect(dynflowLink).not.toHaveClass('disabled');
    });

    it('disables resume and cancel buttons when canEdit is false', () => {
      render(<TaskButtons {...defaultProps} canEdit={false} />);
      expect(screen.getByRole('button', { name: /resume/i })).toBeDisabled();
      expect(screen.getByRole('button', { name: /cancel/i })).toBeDisabled();
    });

    it('disables resume button when resumable is false', () => {
      render(<TaskButtons {...defaultProps} canEdit resumable={false} />);
      expect(screen.getByRole('button', { name: /resume/i })).toBeDisabled();
    });

    it('disables cancel button when cancellable is false', () => {
      render(<TaskButtons {...defaultProps} canEdit cancellable={false} />);
      expect(screen.getByRole('button', { name: /cancel/i })).toBeDisabled();
    });

    it('disables unlock button when state is not paused', () => {
      render(<TaskButtons {...defaultProps} canEdit state="running" />);
      expect(screen.getByRole('button', { name: /^unlock$/i })).toBeDisabled();
    });

    it('enables unlock button when state is paused and canEdit is true', () => {
      render(<TaskButtons {...defaultProps} canEdit state="paused" />);
      expect(
        screen.getByRole('button', { name: /^unlock$/i })
      ).not.toBeDisabled();
    });

    it('disables force unlock button when state is stopped', () => {
      render(<TaskButtons {...defaultProps} canEdit state="stopped" />);
      expect(
        screen.getByRole('button', { name: /force unlock/i })
      ).toBeDisabled();
    });

    it('enables force unlock button when state is not stopped and canEdit is true', () => {
      render(<TaskButtons {...defaultProps} canEdit state="running" />);
      expect(
        screen.getByRole('button', { name: /force unlock/i })
      ).not.toBeDisabled();
    });

    it('renders parent task button when parentTask is provided', () => {
      render(<TaskButtons {...defaultProps} parentTask="parent-123" />);
      const parentButton = screen.getByRole('link', { name: /parent task/i });
      expect(parentButton).toBeInTheDocument();
      expect(parentButton).toHaveAttribute(
        'href',
        '/foreman_tasks/tasks/parent-123'
      );
    });

    it('does not render parent task button when parentTask is not provided', () => {
      render(<TaskButtons {...defaultProps} />);
      expect(
        screen.queryByRole('link', { name: /parent task/i })
      ).not.toBeInTheDocument();
    });

    it('renders sub tasks button when hasSubTasks is true', () => {
      render(<TaskButtons {...defaultProps} hasSubTasks id="task-123" />);
      const subTasksButton = screen.getByRole('link', { name: /sub tasks/i });
      expect(subTasksButton).toBeInTheDocument();
      expect(subTasksButton).toHaveAttribute(
        'href',
        '/foreman_tasks/tasks/task-123/sub_tasks'
      );
    });

    it('does not render sub tasks button when hasSubTasks is false', () => {
      render(<TaskButtons {...defaultProps} hasSubTasks={false} />);
      expect(
        screen.queryByRole('link', { name: /sub tasks/i })
      ).not.toBeInTheDocument();
    });
  });
  describe('user interactions', () => {
    const cancelTaskRequest = jest.fn();
    const resumeTaskRequest = jest.fn();
    const taskProgressToggle = jest.fn();
    const taskReloadStart = jest.fn();
    const id = 'some-id';
    const action = 'some-action';
    const props = {
      ...defaultProps,
      taskReload: false,
      id,
      action,
      cancelTaskRequest,
      resumeTaskRequest,
      taskProgressToggle,
      taskReloadStart,
      status: STATUS.RESOLVED,
      canEdit: true,
      resumable: true,
      cancellable: true,
      state: 'paused',
    };

    it('calls taskProgressToggle when reload button is clicked', () => {
      render(<TaskButtons {...props} />);
      const reloadButton = screen.getByRole('button', {
        name: /start auto-reloading/i,
      });
      fireEvent.click(reloadButton);
      expect(taskProgressToggle).toHaveBeenCalled();
    });

    it('calls taskReloadStart and resumeTaskRequest when resume button is clicked', () => {
      render(<TaskButtons {...props} />);
      const resumeButton = screen.getByRole('button', { name: /resume/i });
      fireEvent.click(resumeButton);
      expect(taskReloadStart).toHaveBeenCalledWith(id);
      expect(resumeTaskRequest).toHaveBeenCalledWith(id, action);
    });

    it('calls taskReloadStart and cancelTaskRequest when cancel button is clicked', () => {
      render(<TaskButtons {...props} />);
      const cancelButton = screen.getByRole('button', { name: /cancel/i });
      fireEvent.click(cancelButton);
      expect(taskReloadStart).toHaveBeenCalledWith(id);
      expect(cancelTaskRequest).toHaveBeenCalledWith(id, action);
    });

    it('calls setUnlockModalOpen when unlock button is clicked', () => {
      render(<TaskButtons {...props} />);
      const unlockButton = screen.getByRole('button', { name: /^unlock$/i });
      fireEvent.click(unlockButton);
      expect(setUnlockModalOpen).toHaveBeenCalledWith(true);
    });

    it('calls setForceUnlockModalOpen when force unlock button is clicked', () => {
      render(<TaskButtons {...props} />);
      const forceUnlockButton = screen.getByRole('button', {
        name: /force unlock/i,
      });
      fireEvent.click(forceUnlockButton);
      expect(setForceUnlockModalOpen).toHaveBeenCalledWith(true);
    });
  });
});
