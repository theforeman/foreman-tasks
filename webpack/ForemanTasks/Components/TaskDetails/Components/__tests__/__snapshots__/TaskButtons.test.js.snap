// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task rendering render with minimal Props 1`] = `
<Col
  bsClass="col"
  componentClass="div"
  xs={12}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="reload-button"
    disabled={false}
    onClick={[MockFunction]}
  >
    <span
      className="glyphicon glyphicon-refresh "
    />
    Start auto-reloading
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="dynflow-button"
    disabled={true}
    href="/foreman_tasks/dynflow/"
    rel="noopener noreferrer"
    target="_blank"
  >
    <span
      data-original-title="dynflow_enable_console Setting is off"
      title="dynflow_enable_console Setting is off"
    >
      Dynflow console
    </span>
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="resume-button"
    data-original-title="You do not have permission"
    disabled={true}
    onClick={[Function]}
    title="You do not have permission"
  >
    Resume
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="cancel-button"
    data-original-title="You do not have permission"
    disabled={true}
    onClick={[Function]}
    title="You do not have permission"
  >
    Cancel
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="unlock-button"
    data-original-title="You do not have permission"
    disabled={true}
    onClick={[MockFunction]}
    title="You do not have permission"
  >
    Unlock
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="force-unlock-button"
    data-original-title="You do not have permission"
    disabled={true}
    onClick={[MockFunction]}
    title="You do not have permission"
  >
    Force Unlock
  </Button>
</Col>
`;

exports[`Task rendering render with some Props 1`] = `
<Col
  bsClass="col"
  componentClass="div"
  xs={12}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="reload-button"
    disabled={false}
    onClick={[MockFunction]}
  >
    <span
      className="glyphicon glyphicon-refresh spin"
    />
    Stop auto-reloading
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="dynflow-button"
    disabled={false}
    href="/foreman_tasks/dynflow/"
    rel="noopener noreferrer"
    target="_blank"
  >
    <span>
      Dynflow console
    </span>
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="resume-button"
    disabled={true}
    onClick={[Function]}
  >
    Resume
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="cancel-button"
    disabled={true}
    onClick={[Function]}
  >
    Cancel
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="parent-button"
    disabled={false}
    href="/foreman_tasks/tasks/parent-id"
  >
    Parent task
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="subtask-button"
    disabled={false}
    href="/foreman_tasks/tasks/test/sub_tasks"
  >
    Sub tasks
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="unlock-button"
    disabled={false}
    onClick={[MockFunction]}
  >
    Unlock
  </Button>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="small"
    bsStyle="default"
    className="force-unlock-button"
    disabled={false}
    onClick={[MockFunction]}
  >
    Force Unlock
  </Button>
</Col>
`;
